src/Npgsql/Internal/NpgsqlConnector.cs

```diff
@@ -1707,32 +1718,59 @@ internal void PerformUserCancellation()
             if (!ReadingPrependedMessagesMRE.Wait(0))
                 return;
 
-            _userCancellationRequested = true;
-
-            if (AttemptPostgresCancellation && SupportsPostgresCancellation)
-            {
-                var cancellationTimeout = Settings.CancellationTimeout;
-                if (PerformPostgresCancellation() && cancellationTimeout >= 0)
-                {
-                    if (cancellationTimeout > 0)
-                    {
-                        ReadBuffer.Timeout = TimeSpan.FromMilliseconds(cancellationTimeout);
-                        ReadBuffer.Cts.CancelAfter(cancellationTimeout);
-                    }
+            PerformUserCancellation();
+        }
+        finally
+        {
+            Monitor.Exit(CancelLock);
+        }
+    }
 
-                    return;
-                }
-            }
+    void PerformDelayedUserCancellation()
```

Comment:
Can this method be merged with PerformImmediateUserCancellation? Or is there some reason we can't take the SyncObj lock in the immediate case?

Reply:
I wish I could! These methods have 2 main differences:

1. `PerformImmediateUserCancellation` is the method called by users for query cancellation (either via cancellation token or NpgsqlCommand.Cancel) which is why it has to check whether it has a bound connection, in multiplexing mode or it has already been called. `PerformDelayedUserCancellation` doesn't do that because it's triggered if and only if `PerformImmediateUserCancellation` has already been called before.
2. `PerformImmediateUserCancellation` also sets `UserCancellationRequested` flag and tries to send the cancellation request immediately by checking whether `ReadingPrependedMessagesMRE` has already been set. `PerformDelayedUserCancellation` doesn't do that because it's called after `ReadingPrependedMessagesMRE` is set.

So yeah, they are similar, but they differ quite a bit at the start and in the middle.